autoload/OmniSharp/util.vim

```diff
@@ -275,7 +275,23 @@ function! OmniSharp#util#TranslatePathForClient(filename) abort
   if filename =~# '\$metadata\$'
     let filename = OmniSharp#util#TempDir() . '/' . fnamemodify(filename, ':t')
   endif
-  return fnamemodify(filename, ':.')
+
+  let modifiers = get(g:, 'OmniSharp_filename_modifiers', ':.')
+
+  if modifiers ==# 'relative'
+    let common = escape(getcwd(), '\')
+    let result = ''
+    while substitute(filename, common . s:dir_separator, '', '') ==# filename
```

Comment:
What about:
```
while substitute(filename, '^' . common . s:dir_separator, '', '') !=# filename
```
- the `'^'` so we only match the beginning
- the `!=#` because we want to continue *until* we've found a common prefix (so as long as we don't match, this is not a prefix and we keep going).

I also think that maybe this should start with
```
let filename = fnamemodify(filename, ':p')
```
Then we're matching on the absolute path, and the `if` below is unnecessary, this works even if the "common" subpath is just `/`.

Reply:
> ...maybe this should start with
> 
> ```vim
> let filename = fnamemodify(filename, ':p')
> ```

I'm not sure it's necessary, as paths received from the server should always be absolute. But it doesn't hurt.

Reply:
> the !=# because we want to continue until we've found a common prefix (so as long as we don't match, this is not a prefix and we keep going).

I'm not quite following you here. Yes, the logic is to continue _until_ we've found a common prefix. As long as the substituted string is the same as `filename`, it means that we have _not_ found a prefix, because the substituted string has not had a prefix removed from it.

Reply:
Sorry, the `!=#` is my bad, got confused by the double negatives :)

About the `':p'`, you're probably right. I got those sometimes relative paths, which was a bit confusing, but I guess that's due to your `':.'`. Probably still doesn't hurt.

It seems to me like the default should probably be either `':p'` (easier for scripting) or your `'relative'` (easier for selecting from a list), but not `':.'`, this seems to me like the worst of both worlds.

Perhaps it should be possible to get different behaviour based on scopes? As I wrote above, I would probably wish for different behaviours when scripting and when displaying. Maybe this function should always return absolute, since it's easier to get from there to anything else, and your `'relative'` should become a method which is called by the receivers, such as the default handler for `Find` which displays the results?

Reply:
The default _should_ be `':p'`, yes. However, that is not backwards-compatible and will give a bad experience for e.g. FZF users. `'relative'` and `':.'` are quite similar, except that `'relative'` goes further than Vim does, in building relative paths. I'm hesitant to make it the default, better to keep things idiomatic.

Making the default `':p'` and applying `'relative'` later is a possibiliy, as long as we can be sure we apply it correctly _everywhere_. Which is tricky. Where should it be applied?

We can't apply it as the default handler to `OmniSharp#actions#usages#Find()`: when a callback is passed to `Find()` it is used _instead_ of using a selector (e.g. FZF). We could perhaps do it by updating the `a:locations` property in [s:CBFindUsages()](https://github.com/OmniSharp/omnisharp-vim/blob/fb44ec4bb108d0755e9ccc2065747496e13fc62e/autoload/OmniSharp/actions/usages.vim#L42), but we then need to ensure we _don't_ do that if a callback exists (or we're back where we started). This is an option. Is that perhaps what you meant by "default handler for `Find`"?

Of course there are also other places this needs to be applied, e.g. `:OmniSharpFindSymbols`.

Reply:
Hi, sorry for disappearing! This is indeed what I meant; I called it d"efault handler" because I had different code organization in mind - similar to the one in my PR, where the top level function (Find) dispatches to the user-provided callback, or to the default one (`s:CBFindUsages`), and then the latter can perform path manipulation, doesn't need to care about callbacks, etc. I will try to send a PR on top of yours to show what I mean.

Regarding the other places where this needs to be applied, I don't know what they all are. I can take a look and try to make a list. I would say, in general it's probably best to leave the scripting interface with absolute paths, and convert only directly before displaying to the user.